### ISSUE TITLE
Spec: learned Harness policy vs static workflow

### ISSUE THREAD

**Issue by TimCole666** [TimCole666/neuroharness#11]
## Problem Statement

PR #10 established the split-neural infrastructure, canonical trajectories, provider-neutral teacher seam, deterministic runtime, and an honest negative result for the closed-set TinyLinearTheta on unseen patch targets.

The next research question should therefore move away from expanding Tiny theta and isolate the part of NeuroHarness that is most distinct from a conventional agent stack: **learned inference-time control inside an otherwise deterministic software Harness**.

Traditional agent Harnesses are software. Their control behavior is usually encoded as some combination of ordinary code, heuristics, system prompts, `AGENTS.md`, or reusable Skills. The current NeuroHarness phi can imitate a hand-written oracle, but the existing six-boolean fixture produces an almost fixed path (`DISCOVER -> TEST -> READ -> THINK -> VERIFY -> STOP`), so there is little room for a learned policy to outperform a good static workflow.

This spec asks a narrower and falsifiable question:

> **Given the same strong external reasoning model, deterministic runtime, tools, verifier, and coding tasks, can a learned local Harness policy outperform a hand-written static workflow in task success or time-to-verified-solution?**

The null hypothesis is equally important:

> **If a static Skill-like workflow performs equally well or better, learned orchestration is unnecessary for this task distribution.**

Negative or null results are valid outcomes.

## Research Goal

Build the smallest controlled experiment that compares three control strategies while holding the external reasoning capability fixed:

1. **Static workflow baseline** — a versioned hand-written policy representing a good software/Skill-style Harness.
2. **Learned Harness policy** — phi trained only from training trajectories and used for held-out control decisions.
3. **LLM-centric control baseline** — the same strong external model is allowed to make the comparable control decisions autoregressively.

All three conditions must use the same deterministic execution and verification kernel and the same external reasoning model for actual code diagnosis/patch generation.

The experiment should answer whether learning the Harness control policy provides measurable value beyond a well-written static workflow.

## Primary Metrics

The primary metrics are deliberately small:

- **Task success rate** — final environment verification, not self-reported completion.
- **Time To Verified Solution (TTVS)** — wall-clock time from task start to verified success; failures remain failures rather than being assigned a fake success time.
- **External model calls** — especially serial calls on the critical path.
- **Harness actions** — total high-level control actions, including redundant/repeated actions.

Secondary diagnostics may include external-model latency, token usage, local compute time, verification failures, retries, and failure modes, but they must not replace the primary metrics.

API dollar cost is not a primary optimization target in this phase.

## Experimental Invariants

To make attribution possible, hold these constant across the three conditions:

- same task split;
- same project fixtures/repositories;
- same deterministic runtime and filesystem safety rules;
- same test/verification commands;
- same external reasoning model and model configuration;
- same patch application contract;
- same maximum run budget / termination rules where applicable;
- same measurement and trajectory schema.

The reference external model may be GPT-5.6-Luna or another configured strong coding model, but the experiment must be provider-neutral and **the exact same model/configuration must be used across compared conditions**.

Model/provider/version, reasoning effort, request parameters, timestamp, latency, and relevant context sent should be recorded when available so later model drift does not get confused with Harness improvement.

## Static Workflow Baseline

The static baseline should be a serious baseline, not a deliberately weak strawman.

It should encode a reasonable coding workflow using versioned software rules and/or concise Skill-like instructions, for example:

- inspect repository when needed;
- run tests when needed;
- gather relevant context;
- invoke the external reasoner for diagnosis/patch generation;
- verify changes;
- retry only when the observed state justifies it;
- stop only after deterministic verification.

The baseline should be inspectable and reproducible. It may be expressible as ordinary code, a Skill, or both, but the benchmark implementation must have deterministic observable control behavior so it can be measured against phi fairly.

Do not intentionally handicap it to make the learned policy win.

## Learned Policy Condition

The learned condition keeps the deterministic Harness software intact and replaces only uncertain control decisions with phi.

Do **not** neuralize invariants such as:

- project-root safety;
- JSON/schema validation;
- process timeouts;
- test exit-code interpretation;
- patch application mechanics;
- mandatory final verification.

Those remain ordinary software.

Phi should learn only decisions for which the correct action genuinely depends on observed state and prior trajectory.

Training data must come only from the training split. Held-out tasks and their trajectories must not influence fitting, hyperparameter selection, threshold tuning, or prompt/rule changes.

## Create Real Control Uncertainty

The current #10 calculator path is useful as a smoke test but is too deterministic for this experiment.

Define a small benchmark where different tasks legitimately require different control trajectories. Examples of useful variation include:

- some failures are immediately localized while others require additional context gathering;
- some candidate patches verify on the first attempt while others require another diagnose/repair cycle;
- some tasks need repository discovery while others expose enough initial structure to skip redundant discovery;
- some contexts are sufficient after one read while others justify additional context before invoking the strong model;
- some verification failures should trigger another reasoning step while others should terminate as explicit failure.

Do not create complexity merely by adding many semantic action labels. Prefer richer observable state and genuine environment variation over hand-designing a large action taxonomy.

The benchmark must contain at least two distinct control-trajectory families in training and at least two in held-out evaluation. A learned policy should not be able to achieve perfect evaluation merely by replaying one fixed six-step sequence.

## Action / State Design

Preserve the principle that phi owns high-level control while the deterministic runtime owns execution.

The existing high-level vocabulary (`DISCOVER`, `TEST`, `READ`, `THINK`, `VERIFY`, `STOP`) may be retained if it can express the required task variation. If retries or repeated reads require a small state/runtime change, make that change explicit and versioned rather than silently bypassing loop detection.

Do not expand the action space unless the benchmark demonstrates that the current vocabulary cannot represent a necessary decision.

The state representation should expose only information that would actually be available at inference time. Avoid benchmark labels, expected patches, hidden fixture names, or other leakage.

## LLM-Centric Baseline

The LLM-centric baseline uses the same strong external model but lets it make the comparable high-level control decisions through language/tool reasoning rather than phi/static rules.

It must use the same runtime/verifier and must not receive privileged information unavailable to the other conditions.

Record the number and latency of serial model calls so the experiment can test whether learned/static control avoids low-value autoregressive steps.

## Tiny Theta

TinyLinearTheta is **not on the critical path of this experiment**.

Do not enlarge it merely to obtain a positive result.

For fair comparison, actual code diagnosis/patch generation should use the same strong external model across the main three conditions. The existing theta implementation remains useful as a prior experiment and optional component, but this spec is about Harness control, not local code-generation capacity.

## Canonical Trajectories

Continue using the canonical `TrajectoryStep` schema introduced and repaired in #10.

Each run should make it possible to recover at least:

- unique `run_id`;
- deterministic case/input fingerprint;
- task identity;
- observed state;
- selected action;
- observation;
- external model/provider/version/configuration when called;
- relevant context sent to the model;
- action/model timing;
- verification result;
- reward/outcome;
- final task status;
- step ordering.

Do not introduce a second trajectory format.

## Train / Held-Out Split

Define an explicit task split before fitting phi.

Requirements:

- train and held-out task fingerprints are disjoint;
- held-out trajectories are never used for training;
- task families are diverse enough that one fixed control sequence is not sufficient;
- split, seeds/configuration, and benchmark version are serialized in experiment output;
- repeated runs of the same case have unique run IDs but stable case fingerprints.

If a small fixture set is used initially, describe it as a controlled benchmark rather than evidence of general coding-agent performance.

## Evaluation

For each held-out task, run all three conditions under the same experiment configuration:

### A. Static workflow

`static policy + same strong model + same runtime/verifier`

### B. Learned Harness

`learned phi + same strong model + same runtime/verifier`

### C. LLM-centric control

`strong model controls comparable decisions + same runtime/verifier`

Report per-task results and aggregates. Do not hide failed tasks or collapse different failure modes into a single success-looking metric.

Where external-model nondeterminism is material, run repeated trials or otherwise record enough configuration to distinguish Harness variance from model variance.

## Interpretation Rules

A positive result requires more than fewer model calls.

Examples:

- If learned phi reduces calls but lowers task success, that is not a win.
- If learned phi matches static success but improves TTVS materially, that is useful evidence.
- If static workflow matches or beats learned phi, report the null/negative result and treat the static workflow as the preferred engineering solution for that task distribution.
- If LLM-centric control wins on novel tasks but loses TTVS on familiar tasks, preserve that tradeoff rather than forcing one global winner.

Do not claim broad coding intelligence from a tiny synthetic benchmark.

## Acceptance Criteria

- [ ] PR #10 remains the frozen prior milestone; this work does not rewrite its negative theta result.
- [ ] A versioned benchmark contains multiple genuine control-trajectory families rather than one fixed six-step path.
- [ ] A strong static Skill/software workflow baseline is implemented and documented.
- [ ] A learned phi condition uses only training trajectories for fitting.
- [ ] An LLM-centric control baseline uses the same external reasoning model and deterministic runtime.
- [ ] All three conditions use the same task split, verifier, patch contract, and external reasoning model/configuration.
- [ ] Held-out task fingerprints and trajectories are excluded from training/tuning.
- [ ] Canonical trajectories record enough provenance to attribute external-model calls and outcomes.
- [ ] Evaluation reports task success, TTVS, external-model calls, and Harness actions per condition.
- [ ] Serial external-model inference on the critical path is separately measurable from cheap local computation.
- [ ] Results retain negative/null outcomes and explicit failure modes.
- [ ] The experiment can conclude that static workflow is better; no implementation choice assumes the neural policy must win.
- [ ] CI remains deterministic and does not require real API credentials; real-model benchmark runs may be opt-in outside CI.
- [ ] The experiment output records benchmark version, split, model/provider/version/configuration, and reproducibility fingerprints where available.

## Out of Scope

- expanding TinyLinearTheta into a larger code model;
- Best-of-N / self-consistency search;
- learned sampling-budget allocation;
- JIT workflow synthesis;
- multi-agent orchestration;
- multi-provider expert routing;
- reinforcement learning through tools/environment;
- long-term memory;
- production OS sandboxing;
- optimizing primarily for API dollar cost;
- claiming general coding-agent superiority from the first controlled benchmark.

## Expected Research Outcome

This spec should produce a falsifiable answer to a narrower question than #3:

> **Are any repeated Harness control decisions actually worth learning, once a good static Skill/workflow and the same strong external model are used as baselines?**

If yes, the next work can identify which decisions amortize well into a learned policy.

If no, that is also a useful result: those Harness decisions should remain static software/Skills, and future learning effort should move elsewhere rather than neuralizing them by default.

**Comment by TimCole666** [comment]
## Implementation tickets

The spec is split into six tickets with one short critical path:

- #12 — **Benchmark: create genuine control uncertainty and freeze the split**
- #13 — **Experiment seam: unify strong reasoner, provenance, and TTVS measurement**
- #14 — **Baseline: implement a strong static Skill/software workflow**
- #15 — **Learned condition: train and evaluate phi on the frozen control benchmark**
- #16 — **Baseline: implement LLM-centric high-level control with the same strong model**
- #17 — **Experiment: run and report the three-way held-out Harness comparison**

### Dependency graph

```text
#12 benchmark / frozen split
  ↓
#13 shared reasoner + provenance + TTVS seam
  ↓
  ├── #14 static Skill/software baseline
  ├── #15 learned phi condition
  └── #16 LLM-centric control baseline
          ↓
     #17 three-way held-out experiment
          ↓
       complete #11
```

For explicit native blockers, use the issue-body relationships:

- #13 blocked by #12
- #14 blocked by #12 and #13
- #15 blocked by #12 and #13
- #16 blocked by #12 and #13
- #17 blocked by #14, #15, and #16

The three condition tickets (#14–#16) are intentionally parallel once #12/#13 are complete. Do not expand Tiny theta, Best-of-N, JIT, multi-agent routing, or RL in this spec.
